firstvoices/backend/serializers/story_serializers.py

```diff
@@ -125,6 +128,44 @@ class Meta(SiteContentLinkedTitleSerializer.Meta):
         )
 
 
+class StoryDetailUpdateSerializer(StorySerializer):
+    pages = serializers.PrimaryKeyRelatedField(
+        queryset=StoryPage.objects.all(),
+        allow_null=True,
+        many=True,
+        validators=[SameSite()],
+    )
+
+    def to_representation(self, instance):
+        data = StorySerializer(instance=instance, context=self.context).data
+        return data
+
+    def update(self, instance, validated_data):
+        if "pages" in validated_data:
+            with transaction.atomic():
+                updated_pages = validated_data["pages"]
+                existing_pages = instance.pages.all()
+                temp_story = Story.objects.create(site=instance.site)
+                for page in existing_pages:
+                    if page not in updated_pages:
+                        page.delete()
+                    else:
+                        page.story = temp_story
+                        page.save()
+                for index, page in enumerate(updated_pages):
+                    if page not in existing_pages:
+                        raise serializers.ValidationError(
+                            f"Page with ID {page.id} does not belong to the story."
+                        )
+                    else:
+                        page.ordering = index
+                        page.story = instance
+                        page.save()
+                temp_story.delete()
+
```

Comment:
I would double check this with @gmcauliffe but I think all pages must be created before trying to save the page order via this API.

So I don't think we need to do the temp_story bit, and I think we should validate that all the listed pages are *already* related to the current story instance, rather than setting the instance on them. I think we have 2 possible validation errors-- "Page with ID x does not exist" and "Page with ID x does not belong to this story"

Reply:
Thanks for the feedback! That was my understanding as well. 

The temp_story bit is there because the combination of a story and ordering is unique together so they cannot be updated in place. They are also required fields so we cannot set them to null and then update them after.

Any existing pages that aren't found in the updated list are deleted so that they aren't left as orphans. I have added a test to ensure this works correctly.

I have refactored this section to use the intersection of the existing and updated pages and I have added comments to make the process clear.

Reply:
That makes sense. I think the bulk_update method could be useful for this scenario, so we don't introduce the possibility of orphaning pages if there's an error while saving one of the pages. E.g., 

# validate the list of updated_pages
...

# set new page order on all page instances
for index, page in enumerate(updated_pages):
  page.ordering = index

# save all pages at once
StoryPage.objects.bulk_update(updated_pages, ['ordering'])

Reply:
(works for me locally)

Reply:
I originally tried `bulk_update` but was getting an integrity error due to the unique together constraint. I just tried it locally again and am still getting this error. It seems like `bulk_update` is not updating the pages in a single query in this case, at least for me.

Additionally, when using `bulk_update` the `pre_save` and `post_save` signals will not be sent which I believe are used to update `StoryPage` search indexing.